File: code/backend/php/src/Core/Domain/Story/StoryValidationError.php
<?php

declare(strict_types=1);

namespace Games\Paths\Core\Domain\Story;

/**
 * StoryValidationError - one referential-integrity or domain-rule violation (Step 22).
 */
final class StoryValidationError
{
    public function __construct(
        public readonly string $rule,
        public readonly string $entityType,
        public readonly ?string $entityId,
        public readonly ?string $field,
        public readonly string $message
    ) {
    }

    public function toArray(): array
    {
        return [
            'rule' => $this->rule,
            'entityType' => $this->entityType,
            'entityId' => $this->entityId,
            'field' => $this->field,
            'message' => $this->message,
        ];
    }
}

File: code/backend/java/core/src/main/java/games/paths/core/model/story/StoryValidationError.java
package games.paths.core.model.story;

/**
 * StoryValidationError - a single referential-integrity or domain-rule violation
 * found by the {@code StoryValidator} (Step 22).
 *
 * @param rule       the rule code that produced the error (e.g. {@code R3_EVENT_REF})
 * @param entityType the story entity type (e.g. {@code choices}, {@code events})
 * @param entityId   identifier of the offending entity within the story (may be null)
 * @param field      the field that holds the broken reference (may be null)
 * @param message    a human-readable description of the violation
 */
public record StoryValidationError(
        String rule,
        String entityType,
        String entityId,
        String field,
        String message) {
}
